hudi-client/hudi-java-client/src/test/java/org/apache/hudi/hadoop/TestHoodieFileGroupReaderOnHive.java

```diff
@@ -0,0 +1,334 @@
+          assertNotNull(compVal);
+          assertArrayWritableEqual(schema, value, compVal);
+        }
+        key = reader.createKey();
+        value = reader.createValue();
+      }
+      reader.close();
+      assertEquals(0, recordMap.size());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private RecordReader<NullWritable, ArrayWritable> createRecordReader(String tablePath) throws IOException {
+    JobConf jobConf = new JobConf(baseJobConf);
+    jobConf.set(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "false");
+
+    TableDesc tblDesc = Utilities.defaultTd;
+    // Set the input format
+    tblDesc.setInputFileFormatClass(HoodieParquetRealtimeInputFormat.class);
+    LinkedHashMap<Path, PartitionDesc> pt = new LinkedHashMap<>();
+    LinkedHashMap<Path, ArrayList<String>> talias = new LinkedHashMap<>();
+
+    PartitionDesc partDesc = new PartitionDesc(tblDesc, null);
+
+    pt.put(new Path(tablePath), partDesc);
+
+    ArrayList<String> arrayList = new ArrayList<>();
+    arrayList.add(tablePath);
+    talias.put(new Path(tablePath), arrayList);
+
+    MapredWork mrwork = new MapredWork();
+    mrwork.getMapWork().setPathToPartitionInfo(pt);
+    mrwork.getMapWork().setPathToAliases(talias);
+
+    Path mapWorkPath = new Path(tablePath);
+    Utilities.setMapRedWork(jobConf, mrwork, mapWorkPath);
+
+    // Add three partition path to InputPaths
+    Path[] partitionDirArray = new Path[HoodieTestDataGenerator.DEFAULT_PARTITION_PATHS.length];
+    Arrays.stream(HoodieTestDataGenerator.DEFAULT_PARTITION_PATHS).map(s -> new Path(tablePath, s)).collect(Collectors.toList()).toArray(partitionDirArray);
+    FileInputFormat.setInputPaths(jobConf, partitionDirArray);
+    jobConf.set(HAS_MAP_WORK, "true");
+    // The following config tells Hive to choose ExecMapper to read the MAP_WORK
+    jobConf.set(MAPRED_MAPPER_CLASS, ExecMapper.class.getName());
+    // setting the split size to be 3 to create one split for 3 file groups
+    jobConf.set(org.apache.hadoop.mapreduce.lib.input.FileInputFormat.SPLIT_MAXSIZE, "128000000");
+    setupJobconf(jobConf);
+
+    HoodieCombineHiveInputFormat combineHiveInputFormat = new HoodieCombineHiveInputFormat();
+    InputSplit[] splits = combineHiveInputFormat.getSplits(jobConf, 1);
+
+    assertEquals(1, splits.length);
+    return  combineHiveInputFormat.getRecordReader(splits[0], jobConf, Reporter.NULL);
+  }
+
+  public void assertArrayWritableEqual(Schema schema, ArrayWritable expected, ArrayWritable actual) {
+    for (Schema.Field field : schema.getFields()) {
+      switch (HoodieAvroUtils.getActualSchemaFromUnion(field.schema(), null).getType()) {
+        case UNION:
+          throw new IllegalStateException("we resolve unions so we should never be here");
+        case RECORD:
+        case MAP:
+        case ENUM:
+        case ARRAY:
+          //TODO: validate complex types
```

Comment:
Let's add a JIRA to track the fix to this.  Essentially, we need to add custom logic to compare complex types, which is important.